library/core/src/ptr/mod.rs

```diff
@@ -436,23 +436,41 @@ mod mut_ptr;
 ///
 /// # Safety
 ///
-/// Behavior is undefined if any of the following conditions are violated:
+/// Immediately upon executing, `drop_in_place` takes out a mutable borrow on the
+/// pointed-to-value. Effectively, this function is implemented like so:
+///
+/// ```
+/// # struct Foo { x: i32 }
+/// unsafe fn drop_in_place(to_drop: *mut Foo) {
+///     let mut value = &mut *to_drop;
+///     // ... drop the fields of `value` ...
+/// }
+/// ```
```

Comment:
What we actually do is stronger than that: we have an `&mut` *function argument*. That makes a different because scope matters.

So if you want to write this in code, I'd suggest something like
````
/// ```
/// # struct Foo { x: i32 }
/// unsafe fn drop_in_place(to_drop: *mut Foo) {
///     drop_in_place_inner(&mut *to_drop);
///     unsafe fn drop_in_place_inner(oto_drop: &mut Foo) {
///         // ... drop the fields of `value` ...
///     }
/// }
/// ```
````